refactor(cli): Improve CLI commands with better error handling and output formatting

Brijeshthummar02 opened the pull request:
#### **What changes were proposed in this pull request?**

This PR enhances multiple CLI commands by improving error handling, resource management, and output formatting. The following changes were made:

-   Improved exception handling for better debugging and user feedback.
-   Enhanced error messages for missing entities to provide clearer guidance.
-   Implemented try-with-resources for better resource management where applicable.
-   Ensured proper handling of empty responses and formatted outputs for better readability.
-   Standardized approach to handling CLI command execution errors.

#### **Why are the changes needed?**

These improvements ensure that CLI commands provide more reliable feedback to users, prevent unhandled exceptions, and improve resource management. The changes also enhance maintainability and debugging for developers.

Fix: #6528 

#### **Does this PR introduce *any* user-facing change?**

Yes, the following user-facing changes are introduced:

-   Improved error messages for missing entities.
-   Better formatting for command outputs.
-   Safer resource handling to prevent unexpected failures.

Brijeshthummar02 commented:
@yuqi1129 need your review on this PR.

yuqi1129 commented:
> @yuqi1129 need your review on this PR.

Got

Brijeshthummar02 commented:
@yuqi1129 That was too quick, Thanks!!

Brijeshthummar02 commented:
@yuqi1129 need your review on this PR again

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/CatalogAudit.java:
```diff
@@ -53,12 +53,20 @@ public void handle() {
       result = client.loadCatalog(this.catalog);
     } catch (NoSuchMetalakeException err) {
       exitWithError(ErrorMessages.UNKNOWN_METALAKE);
+      return; // Stop execution after handling error
```
There is no need for a return here as exitWithEror exits the program.

justinmclean replied:
this applied to all of the added returns below

tengqm replied:
Well ... it is kind of dead code, because this line will never be reached.
However, adding a `return` here does make it more evident?
WDYT, @justinmclean ?

justinmclean replied:
It just adds extra lines to the code in many places, IMO it shouldn't be added. It's also makes it inconsistent with other files, and many other files don't do this - around 130 in total.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/CatalogAudit.java:
```diff
@@ -53,12 +53,20 @@ public void handle() {
+      displayAuditInfo(result.auditInfo());
+    } else {
+      exitWithError("Audit information is not available for this catalog.");
+    }
```
This is no need for an exitWithError here, having no output is fine and not an error.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/CatalogAudit.java:
```diff
@@ -53,12 +53,20 @@ public void handle() {
 
-    displayAuditInfo(result.auditInfo());
+    // **Null check before accessing auditInfo()**
+    if (result != null && result.auditInfo() != null) {
```
only result needs to be checked

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListCatalogProperties.java:
```diff
@@ -50,18 +50,30 @@ public ListCatalogProperties(CommandContext context, String metalake, String cat
+      return;
+    }
+
+    if (gCatalog == null) { // Null check before accessing properties
```
there is no need for comments like "// Null check before accessing properties"

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListCatalogProperties.java:
```diff
@@ -50,18 +50,30 @@ public ListCatalogProperties(CommandContext context, String metalake, String cat
     }
 
     Map<String, String> properties = gCatalog.properties();
+    if (properties == null || properties.isEmpty()) {
```
There is no need for this check and not having properties is not an error.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListColumns.java:
```diff
@@ -59,10 +59,42 @@ public void handle() {
     }
 
-    printResults(columns);
+    // Null / empty check before processing columns
```
The printResults call should be left in

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListCatalogProperties.java:
```diff
@@ -50,18 +50,30 @@ public ListCatalogProperties(CommandContext context, String metalake, String cat
-    try {
-      GravitinoClient client = buildClient(metalake);
+
+    try (GravitinoClient client = buildClient(metalake)) { // Ensures client is closed
```
This i think may end up closing the connection too soon to get the properties

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListFilesetProperties.java:
```diff
@@ -59,21 +60,40 @@ public ListFilesetProperties(
   @Override
   public void handle() {
     Fileset gFileset = null;
+    NameIdentifier name = null;
```
why changethe scope of this?

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListMetalakeProperties.java:
```diff
@@ -45,18 +45,20 @@ public ListMetalakeProperties(CommandContext context, String metalake) {
+      }
+
+      Map<String, String> properties = gMetalake.properties();
+      printProperties(properties);
```
Please leave these lines where they were

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListSchemaProperties.java:
```diff
@@ -54,10 +54,16 @@ public ListSchemaProperties(
+      }
+
+      Map<String, String> properties = gSchema.properties();
+      printProperties(properties);
```
please move the line back to wher ethey were adn do the null check outside of the try block

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListTopicProperties.java:
```diff
@@ -65,16 +65,28 @@ public void handle() {
+      return;
     } catch (Exception exp) {
-      exitWithError(exp.getMessage());
+      exitWithError("An unexpected error occurred: " + exp.getMessage());
```
pleasae don't change this

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/TableAudit.java:
```diff
@@ -48,15 +53,26 @@ public TableAudit(
+    } catch (NoSuchSchemaException err) {
+      exitWithError(ErrorMessages.UNKNOWN_SCHEMA);
+    } catch (NoSuchTableException err) {
+      exitWithError(ErrorMessages.UNKNOWN_TABLE);
```
Are you sure loadTable can throw these exceptions?

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/TopicDetails.java:
```diff
@@ -58,24 +58,27 @@ public TopicDetails(
-    } catch (Exception exp) {
-      exitWithError(exp.getMessage());
+    } catch (Exception e) {
+      exitWithError("Error retrieving topic details: " + e.getMessage());
```
Please don't make unnessacary changes to the code.

tengqm commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListColumns.java:
```diff
@@ -59,10 +59,42 @@ public void handle() {
+    }
+
+    StringBuilder all = new StringBuilder();
+    all.append("name,datatype,comment,nullable,auto_increment").append(System.lineSeparator());
```
I believe we had a conversation over the "autoIncrement" field ...
It is not supposed appear on every column.

Brijeshthummar02 commented:
@justinmclean @tengqm tried to fix and resolve as suggested can just go through it ones also removed what i feel is not needed.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/CatalogAudit.java:
```diff
@@ -53,20 +53,15 @@ public void handle() {
 
-    // **Null check before accessing auditInfo()**
-    if (result != null && result.auditInfo() != null) {
+    // Only check if result is null
```
Please remove this comment, there is no need for comments that explain simple code. They don't really add anything and at worse get out of sync with the code and end up causing confusion. See for instance https://stackoverflow.blog/2021/12/23/best-practices-for-writing-code-comments/

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListCatalogProperties.java:
```diff
@@ -49,31 +48,19 @@ public ListCatalogProperties(CommandContext context, String metalake, String cat
-      gCatalog = client.loadCatalog(catalog);
+      if (gCatalog != null) {
+          printProperties(gCatalog.properties());
+      }
```
Please move this out of the try block, adding it here makes it unclear what call throws what exception.

justinmclean replied:
This has not been reloved.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListColumns.java:
```diff
@@ -48,53 +47,57 @@ public ListColumns(
-    }
+                    if (hasAutoIncrement) {
+                        columnDetails.append(",").append(column.autoIncrement() ? "true" : "");
+                    }
```
While what you are doing here seems like a good idea, it can break any scripts that users have already written. Having a column that can optionally show up will also make it harder for scripts to use this output.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListFilesetProperties.java:
```diff
@@ -60,27 +60,28 @@ public ListFilesetProperties(
-      gFileset = client.loadCatalog(catalog).asFilesetCatalog().loadFileset(name);
+        NameIdentifier name = NameIdentifier.of(schema, fileset);
+
+        try (GravitinoClient client = buildClient(metalake)) {
```
as I mentione dbefore that may close the client before we can make other  needed calls

justinmclean commented:
The CI is failing as we use spotless to make sure the code has consistent formatting. You'll need to run `./gradlew :clients:cli:spotlessApply` and check in the changes to fix this.

Brijeshthummar02 commented:
@justinmclean i can not run ./gradlew :clients:cli:spotlessApply as i am using java 21 and it req java 17.

Brijeshthummar02 commented:
@justinmclean as you see build is successful.

![Screenshot 2025-03-02 114325](https://github.com/user-attachments/assets/d2b4a325-3577-422f-8dd3-ca49ee8be166)

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/CreateTable.java:
```diff
@@ -82,27 +82,47 @@ public void handle() {
+      return; // Stop execution
     } catch (Exception exp) {
       exitWithError("Error initializing client or table name: " + exp.getMessage());
+      return;
```
Please remove the unneeded return statements. This is inconsistent with other files here.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/CreateTable.java:
```diff
@@ -82,27 +82,47 @@ public void handle() {
+      if (tableName == null) {
+        exitWithError("Table name could not be determined.");
+        return;
+      }
```
I don't think it is possible for the table name to be null, options passing in on the command line cannot be null, but I guess it doesn't hurt. There's also no real need for it to be in the try block.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListColumns.java:
```diff
@@ -51,18 +50,42 @@ public ListColumns(
+              + Joiner.on(".").join(metalake, catalog, schema, table));
     } catch (Exception exp) {
-      exitWithError(exp.getMessage());
+      exitWithError("An error occurred while retrieving column details: " + exp.getMessage());
```
You need to use printResults to print out this information.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListFilesetProperties.java:
```diff
@@ -59,21 +59,40 @@ public ListFilesetProperties(
+    } finally {
+      if (client != null) {
+        client.close();
+      }
```
This will close the client too soon.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListMetalakeProperties.java:
```diff
@@ -46,17 +46,22 @@ public ListMetalakeProperties(CommandContext context, String metalake) {
-    try {
-      GravitinoAdminClient client = buildAdminClient();
+
+    try (GravitinoAdminClient client = buildAdminClient()) { // Ensure resource cleanup
```
This is likely to close the client too early.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListMetalakeProperties.java:
```diff
@@ -46,17 +46,22 @@ public ListMetalakeProperties(CommandContext context, String metalake) {
+        exitWithError("Metalake not found: " + metalake);
+        return;
+      }
+
```
best to move this outside of the try block.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListSchemaProperties.java:
```diff
@@ -55,8 +55,8 @@ public ListSchemaProperties(
-    try {
-      GravitinoClient client = buildClient(metalake);
+
+    try (GravitinoClient client = buildClient(metalake)) { // Ensure resource cleanup
```
Agin this may not work.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListTableProperties.java:
```diff
@@ -59,11 +59,17 @@ public ListTableProperties(
   public void handle() {
-    Table gTable = null;
-    try {
+    try (GravitinoClient client = buildClient(metalake)) { // Ensures resource cleanup
```
This may not work.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListTableProperties.java:
```diff
@@ -59,11 +59,17 @@ public ListTableProperties(
+      }
+
+      Map<String, String> properties = gTable.properties();
+      printProperties(properties);
```
Please move outside of try block.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListTagProperties.java:
```diff
@@ -53,12 +53,22 @@ public void handle() {
+        exitWithError(ErrorMessages.UNKNOWN_TAG);
+        return;
+      }
+
```
Please move outside of try block

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListTagProperties.java:
```diff
@@ -53,12 +53,22 @@ public void handle() {
     } catch (Exception exp) {
-      exitWithError(exp.getMessage());
+      exitWithError("An unexpected error occurred: " + exp.getMessage());
+      return;
```
Please remove the return calls.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListTopicProperties.java:
```diff
@@ -65,6 +65,13 @@ public void handle() {
+        exitWithError(ErrorMessages.UNKNOWN_TOPIC);
+        return;
+      }
+
```
Please move outside try block and remove unneeded comment

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ModelDetails.java:
```diff
@@ -68,20 +68,34 @@ public void handle() {
+      exitWithError("An unexpected error occurred: " + err.getMessage());
+      return;
     }
+
```
Please refer to the comments above.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/SchemaAudit.java:
```diff
@@ -52,20 +52,23 @@ public SchemaAudit(CommandContext context, String metalake, String catalog, Stri
+      if (result == null) {
+        exitWithError("Failed to retrieve schema details.");
+        return;
+      }
```
Please move outside of try block

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/SchemaAudit.java:
```diff
@@ -52,20 +52,23 @@ public SchemaAudit(CommandContext context, String metalake, String catalog, Stri
       exitWithError(ErrorMessages.UNKNOWN_SCHEMA);
     } catch (Exception exp) {
-      exitWithError(exp.getMessage());
+      exitWithError("An unexpected error occurred: " + exp.getMessage());
```
There no need to change this.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/TableAudit.java:
```diff
@@ -48,15 +52,24 @@ public TableAudit(
+      if (gTable == null) {
+        exitWithError("Failed to retrieve table details.");
+        return;
+      }
```
please move outside of try block

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/TableAudit.java:
```diff
@@ -48,15 +52,24 @@ public TableAudit(
+      exitWithError(ErrorMessages.UNKNOWN_SCHEMA);
     } catch (Exception exp) {
-      exitWithError(exp.getMessage());
+      exitWithError("An unexpected error occurred: " + exp.getMessage());
```
there no need for this change

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/TopicDetails.java:
```diff
@@ -58,12 +58,17 @@ public TopicDetails(
+      }
+
+      printResults("Topic Name: " + gTopic.name() + ", Comment: " + gTopic.comment());
+
```
please move the code to outside the try block

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/UserDetails.java:
```diff
@@ -47,23 +47,21 @@ public UserDetails(CommandContext context, String metalake, String user) {
+        printInformation("The user has no assigned roles.");
+      } else {
+        printResults("User Roles: " + String.join(", ", roles));
+      }
```
please move to outside the try block

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/UserDetails.java:
```diff
@@ -47,23 +47,21 @@ public UserDetails(CommandContext context, String metalake, String user) {
-    } else {
-      printResults(String.join(",", roles));
+    } catch (Exception e) {
+      exitWithError("Error retrieving user details: " + e.getMessage());
```
This error may be misleading, please don't change it.

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ModelDetails.java:
```diff
@@ -68,20 +68,34 @@ public void handle() {
+      if (gModel == null) {
+        exitWithError(ErrorMessages.UNKNOWN_MODEL);
+        return;
+      }
```
please move outside the try block

Brijeshthummar02 commented:
@justinmclean tried to fix all return statement, using it was my best practice but for this repo as u said we don't generally need so removed apart from that fixed all if statements outside the try block. Also thankyouuu! for giving reviews always as there are multiples files so this is taking way more time.

tengqm commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListCatalogProperties.java:
```diff
@@ -49,10 +49,14 @@ public ListCatalogProperties(CommandContext context, String metalake, String cat
-      GravitinoClient client = buildClient(metalake);
-      gCatalog = client.loadCatalog(catalog);
+      Catalog gCatalog = client.loadCatalog(catalog);
     } catch (NoSuchMetalakeException err) {
```
Oh No. This code won't work.
Checking gCatalog on line 54 before it is declared on line 59?
Throw NoSuchMetalakeException while metalake is used on line 52?

tengqm replied:
Will loadCatalog throw NoSuchMetalakeException?

Brijeshthummar02 replied:
On checking loadCatalog it will generally not throw NoSuchMetalakeException, and hence removing it.

tengqm replied:
The original code `buildClient(metalake);` may throw this exception, right?
You moved that line out of the try block for unknown reason.
Now you know why I'm concerned?

tengqm commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListFilesetProperties.java:
```diff
@@ -59,10 +59,15 @@ public ListFilesetProperties(
   @Override
   public void handle() {
     Fileset gFileset = null;
+    GravitinoClient client = null;
```
Why this change?

tengqm replied:
I don't think this was resolved.

tengqm commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/TableAudit.java:
```diff
@@ -53,10 +57,22 @@ public void handle() {
 
+    if (gTable == null) {
+      exitWithError("Failed to retrieve table details.");
+      return;
```
I don't have a strong opinion about this additional `return` line, though I'm more leaning towards having it (no harm + code readability).
However, when we decide to add these `return`s, we need to be consistent.
For example, line 68 doesn't have a following `return` statement, this would confuse others.

Brijeshthummar02 commented:
@tengqm made relevant changes.

tengqm commented:
> @tengqm made relevant changes.

Two points here:

- If you don't agree to a feedback, you can leave it there, but don't mark it as "resolved". Marking a feedback as "resolved" without changing the code accordingly means "I don't care" or "I'm ignoring what you suggested.".
- For the "return" statement, I would say that again. I don't have a strong opinion on whether we should add it. But if there is decision, we need to be consistent in the code.

Brijeshthummar02 commented:
@tengqm Apologies for any confusion. As Justin mentioned, the decision was to remove the return statement, and I’ve tried to remove it from almost all files. I didn’t mean to take your feedback the wrong way—I truly appreciate the review. I might have missed checking all files since I’m currently in the middle of my university exams while still staying active in open-source contributions.

I’ll make sure to correct it properly. Would it be fine if I just remove the return statement from all files to ensure consistency?

Brijeshthummar02 commented:
@tengqm Just to confirm, should I go ahead and remove the return statement from all cases, including if statements?

tengqm commented:
> @tengqm Just to confirm, should I go ahead and remove the return statement from all cases, including if statements?

Yup. If the decision was that 'return' is unnecessary, please remove them **all**.

tengqm commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListColumns.java:
```diff
@@ -51,18 +50,40 @@ public ListColumns(
+            .append(",");
+        all.append(column.comment() != null ? column.comment() : "N/A").append(",");
+        all.append(column.nullable() ? "true" : "false").append(",");
+        all.append(column.autoIncrement() ? "true" : "false");
```
Not sure if this column is okay and consistent with other format of outputs.

Brijeshthummar02 commented:
@tengqm do we need any changes??

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListTables.java:
```diff
@@ -47,43 +47,49 @@ public ListTables(CommandContext context, String metalake, String catalog, Strin
-      printInformation("No tables exist.");
-      return;
-    }
+      if (tables == null || tables.length == 0) {
```
this code should be outside the try block

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ListTagProperties.java:
```diff
@@ -53,12 +53,17 @@ public void handle() {
       exitWithError(ErrorMessages.UNKNOWN_TAG);
     } catch (Exception exp) {
-      exitWithError(exp.getMessage());
+      exitWithError("An unexpected error occurred: " + exp.getMessage());
```
Please revert this change

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/ModelDetails.java:
```diff
@@ -68,20 +68,27 @@ public void handle() {
       exitWithError(ErrorMessages.UNKNOWN_MODEL);
     } catch (Exception err) {
-      exitWithError(err.getMessage());
+      exitWithError("An unexpected error occurred: " + err.getMessage());
```
please revet this change

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/SchemaAudit.java:
```diff
@@ -66,6 +68,11 @@ public void handle() {
 
+    if (result == null) {
+      exitWithError("Failed to retrieve schema details.");
+      return;
```
no need for a return here

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/UserDetails.java:
```diff
@@ -49,21 +49,20 @@ public UserDetails(CommandContext context, String metalake, String user) {
+    try (GravitinoClient client = buildClient(metalake)) {
       roles = client.getUser(user).roles();
-    } catch (NoSuchMetalakeException err) {
+    } catch (NoSuchMetalakeException e) {
```
there is no need to change these names

justinmclean commented on clients/cli/src/main/java/org/apache/gravitino/cli/commands/UserDetails.java:
```diff
@@ -49,21 +49,20 @@ public UserDetails(CommandContext context, String metalake, String user) {
+      printInformation("The user has no assigned roles.");
     } else {
-      printResults(String.join(",", roles));
+      printResults("User Roles: " + String.join(", ", roles));
```
Please don't change the output as this may break things

Brijeshthummar02 commented:
@jerryshao i really appreciate your review on this PR.

justinmclean commented:
@Brijeshthummar02, there are still some outstanding, unaddressed comments. You have also resolved several issues without fixing them. It's getting there, but I think it still needs a little work.

My main concern is that the added `try (GravitinoClient client = buildClient(metalake))` will close the client too soon, our unit tests are unlikely to catch this. I'll take a look on Monday to see if this is an issue.

Brijeshthummar02 commented:
> @Brijeshthummar02, there are still some outstanding, unaddressed comments. You have also resolved several issues without fixing them. It's getting there, but I think it still needs a little work.
> 
> My main concern is that the added `try (GravitinoClient client = buildClient(metalake))` will close the client too soon, our unit tests are unlikely to catch this. I'll take a look on Monday to see if this is an issue.

Follow up
As you mentioned you will take a look and point out issue if exist.
